File: java/platform-data-doublets-client/src/main/java/platform/data/doublets/client/Link.java
package platform.data.doublets.client;

/**
 * Represents a link in the doublets data structure.
 * A link connects two nodes with an identifier.
 */
public class Link {
    private final long id;
    private final long fromId;
    private final long toId;

    /**
     * Creates a new Link instance.
     *
     * @param id     the unique identifier of the link
     * @param fromId the identifier of the source node
     * @param toId   the identifier of the target node
     */
    public Link(long id, long fromId, long toId) {
        this.id = id;
        this.fromId = fromId;
        this.toId = toId;
    }

    /**
     * Gets the unique identifier of this link.
     *
     * @return the link identifier
     */
    public long getId() {
        return id;
    }

    /**
     * Gets the identifier of the source node.
     *
     * @return the source node identifier
     */
    public long getFromId() {
        return fromId;
    }

    /**
     * Gets the identifier of the target node.
     *
     * @return the target node identifier
     */
    public long getToId() {
        return toId;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        Link link = (Link) obj;
        return id == link.id && fromId == link.fromId && toId == link.toId;
    }

    @Override
    public int hashCode() {
        return Long.hashCode(id) * 31 + Long.hashCode(fromId) * 17 + Long.hashCode(toId);
    }

    @Override
    public String toString() {
        return String.format("Link{id=%d, fromId=%d, toId=%d}", id, fromId, toId);
    }
}

File: php/src/Link.php
<?php

declare(strict_types=1);

namespace LinksPlatform\Data\Doublets\Gql;

/**
 * Represents a single Link in the Doublets system.
 * 
 * A Link is a triplet consisting of:
 * - id: The unique identifier of the link
 * - fromId: The source node ID
 * - toId: The target node ID
 */
class Link
{
    private int $id;
    private int $fromId;
    private int $toId;

    public function __construct(int $id, int $fromId, int $toId)
    {
        $this->id = $id;
        $this->fromId = $fromId;
        $this->toId = $toId;
    }

    /**
     * Create a Link from GraphQL response data.
     * 
     * @param array $data Array containing 'id', 'from_id', 'to_id' keys
     * @return self
     */
    public static function fromArray(array $data): self
    {
        return new self(
            (int)$data['id'],
            (int)$data['from_id'],
            (int)$data['to_id']
        );
    }

    /**
     * Get the link ID.
     * 
     * @return int
     */
    public function getId(): int
    {
        return $this->id;
    }

    /**
     * Get the source node ID.
     * 
     * @return int
     */
    public function getFromId(): int
    {
        return $this->fromId;
    }

    /**
     * Get the target node ID.
     * 
     * @return int
     */
    public function getToId(): int
    {
        return $this->toId;
    }

    /**
     * Convert the link to an array representation.
     * 
     * @return array
     */
    public function toArray(): array
    {
        return [
            'id' => $this->id,
            'from_id' => $this->fromId,
            'to_id' => $this->toId,
        ];
    }

    /**
     * Convert the link to a JSON string.
     * 
     * @return string
     */
    public function toJson(): string
    {
        return json_encode($this->toArray(), JSON_THROW_ON_ERROR);
    }

    /**
     * String representation of the link.
     * 
     * @return string
     */
    public function __toString(): string
    {
        return sprintf('Link(id: %d, from: %d, to: %d)', $this->id, $this->fromId, $this->toId);
    }
}